### DISCUSSION TITLE
Extend the list of permitted attribute parameter types

### DISCUSSION THREAD

**Discussion by adejongh93** [dotnet/csharplang#4771] [Language Ideas]
### Background
This code throws compile error

``` csharp
using System;

namespace CSharp_Language_Proposals.AttributeProposal
{
    class CustomAttribute : Attribute
    {
        public CustomAttribute(DateTime param)
        {
        }
    }

    class Class1
    {
        [Custom(DateTime.Now)] // CS0181 Attribute constructor parameter 'param' has type DateTime,
                               // which is not a valid attribute parameter type
        void Method()
        {
        }
    }
} 
```

### Proposal
Modify the compiler so the attributes would accept new types and runtime evaluated expressions.

``` csharp
using System;

namespace CSharp_Language_Proposals.AttributeProposal
{
    class CustomAttribute : Attribute
    {
        public CustomAttribute(DateTime param)
        {
        }
    }

    class Class1
    {
        [Custom(DateTime.Now)] // Compiler builds correctly.
        void Method()
        {
        }
    }
}

**Comment by bernd5** [comment]
What should be the value of "DateTime.Now"? Should it be the time when the binary was build? Should it re-evaluated at runtime to the current time?
Why not implementing such logic in the attribute? To configure it you could use constants, e.g. strings and enums.

**Comment by Unknown6656** [comment]
Mildly related:
- #124 (see https://github.com/dotnet/roslyn/tree/features/generic-attributes)
- #4770
- #343
- https://github.com/dotnet/roslyn/issues/953

There was one other proposal, but I was unable to find it. It was about compile-time constant values for arbitrary blittable/struct types.

**Comment by HaloFour** [comment]
The TL;DR of all of the referenced issues/discussions is that the list of supported types permitted in an attribute's properties and parameters are limited by the runtime.  The runtime would need to be modified to support other types and establish how those types are encoded into the metadata of the assembly.

**Reply by Unknown6656** [reply]
That is indeed true, however I am pretty sure that there was a proposal to enable compile-time constants for arbitrary blittable/struct types. This would allow `System.DateTime` to be used as `const` fields, as default parameter value, **and** as Attribute argument.

(though I am currently unable to find the proposal. maybe it was in dotnet/roslyn)

**Reply by HaloFour** [reply]
That depends on how it's implemented.  Languages can already fake constants, which is why C# can support constants and default parameter values of `System.Decimal`, and why VB.NET can support constants and default parameter values of `System.DateTime`. But neither can be used in attributes because the runtime is not in on it.  So any support for blittable constants would need to have runtime support in order to expand on the parameters allowed in attributes.

Lambdas would be a bit of a different ballgame, though.  I guess technically the method token could be a constant value which could be encoded in the assembly metadata.

**Reply by acaly** [reply]
Limitation of runtime shouldn't be limitation of adding new features. There are many new features added to .NET requiring both sides to work cooperatively.

It is equally valid for the runtime maintainer to say that no language will support this feature so runtime won't add it.
